src/bezier.cpp

```diff
@@ -560,94 +519,100 @@ PointVector Curve::pointsOfIntersection(const Curve& curve, bool stop_at_first,
   return points_of_intersection;
 }
 
-double Curve::projectPoint(const Point& point, double step, double epsilon, std::size_t max_iter) const
+Parameter Curve::projectPoint(const Point& point, double epsilon) const
 {
-  step = std::max(step, 0.01);
-  epsilon = std::max(epsilon, 0.001);
+  /* Chen, X. D., Zhou, Y., Shu, Z., Su, H., & Paul, J. C. (2007, August). Improved Algebraic Algorithm on Point
+   * projection for Bezier curves. In Second International Multi-Symposiums on Computer and Computational Sciences
+   * (IMSCCS 2007) (pp. 158-163). IEEE.
+   */
 
-  double t = 0;
-  double t_dist = (valueAt(t) - point).norm();
-
-  // Coarse search
-  for (double k = step; k < 1 + step; k += step)
+  if (!cached_projection_polynomial_part_)
   {
-    double new_dist = (valueAt(k) - point).norm();
-    if (new_dist < t_dist)
-    {
-      t_dist = new_dist;
-      t = k;
-    }
+    Eigen::MatrixXd curve_polynomial = (bernsteinCoeffs() * control_points_);
+    Eigen::MatrixXd derivate_polynomial = (derivative()->bernsteinCoeffs() * derivative()->control_points_);
+
+    Eigen::VectorXd polynomial_part = Eigen::VectorXd::Zero(curve_polynomial.rows() + derivate_polynomial.rows() - 1);
+    for (uint k = 0; k < curve_polynomial.rows(); k++)
+      for (uint i = 0; i < derivate_polynomial.rows(); i++)
+        polynomial_part(k + i) += curve_polynomial.row(k).dot(derivate_polynomial.row(i));
+
+    const_cast<Curve*>(this)->cached_projection_polynomial_part_.reset(new Eigen::VectorXd(std::move(polynomial_part)));
+    const_cast<Curve*>(this)->cached_projection_polynomial_derivative_ = std::move(derivate_polynomial);
   }
 
-  // Fine search - Halley
-  // function to minimize is a dot product between projection vector and tangent
-  // - projection vector is a vector between point we are projecting and our current guess
-  double t_old = t;
-  std::size_t current_iter = 0;
-  while (current_iter < max_iter)
+  Eigen::VectorXd polynomial = *cached_projection_polynomial_part_;
+  for (uint i = 0; i < cached_projection_polynomial_derivative_.rows(); i++)
+    polynomial(i) -= point.dot(cached_projection_polynomial_derivative_.row(i));
+
+  double projection = (point - valueAt(0.0)).norm() < (point - valueAt(1.0)).norm() ? 0.0 : 1.0;
+  double min = (point - valueAt(projection)).norm();
+
+  for (auto candidate : Sturm::roots(polynomial.reverse(), Sturm::RootTypeFlag::Convex, epsilon))
```

Comment:
Think this could be done with stl

Reply:
yes, but it would result in more code and operations:
- there are two additional candidates to be checked (line 547)
- after the correct candidate is found, dist wold have to be calculated again